### PR TITLE
Add CLI Commands

### LINKED ISSUE #147: Generate Announcement from CLI/OCC
### Steps to reproduce
N/A

### Expected behaviour
As a Request for Enhancement/Feature Request:
As a Nextcloud admin, it would be helpful to be able to create announcements from the command line (for example via Nextcloud's OCC).

### Actual behaviour
Announcements can, to my understanding, only be created from the Announcement Center.

### Server configuration
**Operating system**: Ubuntu 18.04.2 LTS (Bionic Beaver)

**Web server:** Apache 2.4.29

**Database:** MariaDB 10.1.38

**PHP version:** PHP 7.2.15

**Nextcloud version:** 15.0.5

**Where did you install Nextcloud from:** Archive File (see https://nextcloud.com/install/#instructions-server)

**Signing status:** N/A

**List of activated apps:** N/A

**Nextcloud configuration:** N/A

**Are you using an external user-backend, if yes which one:** N/A

### Client configuration
**Browser:** N/A

**Operating system:** N/A

### Logs
N/A

#### Nextcloud log (data/owncloud.log)
N/A

#### Browser log
N/A

### PR DESCRIPTION
**Pull request by mwinkens**
This PR adds CLI commands for announcing, deleting and viewing announcements:

```
announcementcenter:announce [--activities] [--notifications] [--emails] [--comments] [-s|--schedule-time [SCHEDULE-TIME]] [-d|--delete-time [DELETE-TIME]] [-g|--group [GROUP]] [--] <user> <subject> <message>

announcementcenter:delete <id>

announcementcenter:list [<limit>]
```

# Help:
<details>
<summary>CLICK ME</summary>

## Announce:

```
$php occ announcementcenter:announce --help
Description:
  Create an announcement

Usage:
  announcementcenter:announce [options] [--] <user> <subject> <message>

Arguments:
  user                                 User who creates the announcement
  subject                              User for whom the addressbook will be created
  message                              message of the announcement (supports markdown)

Options:
      --activities                     Get notified over activities
      --notifications                  Get notified over nextclouds notifications
      --emails                         Notify users over email
      --comments                       Allow comments
  -s, --schedule-time[=SCHEDULE-TIME]  Publishing time of the announcement (see php strtotime)
  -d, --delete-time[=DELETE-TIME]      Deletion time of the announcement (see php strtotime)
  -g, --group[=GROUP]                  group to set send announcement to (default "everyone", multiple allowed) [default: ["everyone"]] (multiple values allowed)
  -h, --help                           Display help for the given command. When no command is given display help for the list command
  -q, --quiet                          Do not output any message
  -V, --version                        Display this application version
      --ansi|--no-ansi                 Force (or disable --no-ansi) ANSI output
  -n, --no-interaction                 Do not ask any interactive question
      --no-warnings                    Skip global warnings, show command output only
  -v|vv|vvv, --verbose                 Increase the verbosity of messages: 1 for normal output, 2 for more verbose output and 3 for debug
```

## List

```
$php occ announcementcenter:list --help
Description:
  List all announcements

Usage:
  announcementcenter:list [<limit>]

Arguments:
  limit                 Maximal number of announcements listed [default: 10]

Options:
  -h, --help            Display help for the given command. When no command is given display help for the list command
  -q, --quiet           Do not output any message
  -V, --version         Display this application version
      --ansi|--no-ansi  Force (or disable --no-ansi) ANSI output
  -n, --no-interaction  Do not ask any interactive question
      --no-warnings     Skip global warnings, show command output only
  -v|vv|vvv, --verbose  Increase the verbosity of messages: 1 for normal output, 2 for more verbose output and 3 for debug

```

## Delete

```
php occ announcementcenter:delete --help
Description:
  Delete an announcement by id

Usage:
  announcementcenter:delete <id>

Arguments:
  id                    Id of announcement to delete

Options:
  -h, --help            Display help for the given command. When no command is given display help for the list command
  -q, --quiet           Do not output any message
  -V, --version         Display this application version
      --ansi|--no-ansi  Force (or disable --no-ansi) ANSI output
  -n, --no-interaction  Do not ask any interactive question
      --no-warnings     Skip global warnings, show command output only
  -v|vv|vvv, --verbose  Increase the verbosity of messages: 1 for normal output, 2 for more verbose output and 3 for debug

```

</details>

# Examples:
<details>
<summary>CLICK ME</summary>

### Minimal

Minimal, you need at least one announcement option:
```
php occ announcementcenter:announce nextcloud27 'a test' 'a test message'
```
Announces as user nextcloud27 a message ('a test message') with a subject ('a test').
Default group is `everyone`, so this announcement reaches everyone. However this didn't use any notification type, like `--emails` and is just visible in the announcement app

### Groups

You can specify multiple groups with `--group` or `-g`:
```
php occ announcementcenter:announce nextcloud27 'a test' 'a test message' --emails --group group1 --group group2
```
This example only announces to users with group `group1` and `group2` and sends each user in the group an email

### Scheduling/Deletion

You can set the scheduled announcement time with `--schedule-time <time>`, while the time can be a unix timestamp or any string that can be interpreted by [strtotime](https://www.php.net/manual/de/function.strtotime.php)

**Unix-timestamp**:
```
php occ announcementcenter:announce nextcloud27 'a test' 'a test message' --emails --schedule-time 1711443260
```

**String**:
Tomorrow:
```
php occ announcementcenter:announce nextcloud27 'a test' 'a test message' --emails --schedule-time tomorrow
```

Friday:
```
php occ announcementcenter:announce nextcloud27 'a test' 'a test message' --emails --schedule-time friday
```

16:45
```
php occ announcementcenter:announce nextcloud27 'a test' 'a test message' --emails --schedule-time 16:45
```

delete-time is similar

## List
```
$php occ announcementcenter:list
ID  |Subject                 |Message                                           
----|------------------------|--------------------------------------------------
33  |testin a bigger subje...|test                                              
25  |test9                   |test9                                             
24  |Your Subject            |Your Plain Body                                   
23  |test8                   |test8                                             
21  |test6                   |test6                                             
20  |test5                   |test5                                             
19  |test4                   |test4                                             
18  |test3                   |lorem ipsum I don't speak latin                   
17  |test2                   |testm    
```

You can specify the output limit, default is 10:
```
$php occ announcementcenter:list 3
ID  |Subject                 |Message                                           
----|------------------------|--------------------------------------------------
33  |testin a bigger subje...|test                                              
25  |test9                   |test9                                             
24  |Your Subject            |Your Plain Body                                   
And more ...
```

## Delete
Delete announcement with ID 33:
```
$php occ announcementcenter:delete 33
Successfully deleted #33
```

</details>

# Roadmap

- [x] tests!

---

based on #767 (merged)
closes #147

### CODE REVIEW

**Comment by mwinkens** [comment]
@nickvergessen this is also ready for review and if you want we can integrate this into the next update! ~400 lines of this ~800 lines PR are just unit tests, so this shouldn't take too much effort. Also this PR doesn't have a frontend :tada:

File: tests/PHPUnitUtil.php [hunk 1]
```diff
@@ -0,0 +1,12 @@
+<?php
+
+namespace OCA\AnnouncementCenter\Tests;
+
+class PHPUnitUtil {
+	public static function callHiddenMethod($obj, $name, array $args) {
```
Comment (nickvergessen) [inline tests/PHPUnitUtil.php:6]: This exists already as `self::invokePrivate()` from nextcloud side :)

File: psalm.xml [hunk 2]
```diff
@@ -28,6 +28,11 @@
 				<referencedClass name="Doctrine\DBAL\Types\Types" />
 				<referencedClass name="OC" />
 				<referencedClass name="OCA\Guests\UserBackend" />
+				<referencedClass name="Symfony\Component\Console\Command\Command" />
```
Comment (nickvergessen) [inline psalm.xml:31]: Not a fan of this thirdparty access. Symfony version is not the same across all supported versions.
Also by ignoring it in psalm, CI will not detect if there is an issue. But yeah we don't have a better option atm.

File: lib/Command/Announce.php [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+				'activities',
+				null,
+				InputOption::VALUE_NONE,
+				'Get notified over activities',
+			)
+			->addOption(
+				'notifications',
+				null,
+				InputOption::VALUE_NONE,
+				'Get notified over nextclouds notifications',
```
Comment (nickvergessen) [inline lib/Command/Announce.php:80]:
```suggestion
				'Generate notifications',
```

File: lib/Command/Announce.php [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+			)
+			->addOption(
+				'emails',
+				null,
+				InputOption::VALUE_NONE,
+				'Notify users over email',
```
Comment (nickvergessen) [inline lib/Command/Announce.php:86]:
```suggestion
				'Notify users via email',
```

File: lib/Command/Announce.php [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+	private function plainifyMessage(string $message) {
+		# TODO use Parsedown or Markdownify here
+		return $message;
+	}
+
+	protected function execute(InputInterface $input, OutputInterface $output): int {
+		// required
+		$user = $input->getArgument('user');
+		if (!$this->userManager->userExists($user)) {
+			throw new \InvalidArgumentException("User <$user> in unknown.");
```
Comment (nickvergessen) [inline lib/Command/Announce.php:126]:
```suggestion
			$output->writeln("User <$user> in unknown.");
			return 1;
```

File: lib/Command/Announce.php [hunk 6]
```diff
@@ -0,0 +1,197 @@
...
+		$emails = $input->getOption('emails');
+		$comments = $input->getOption('comments');
+
+		// times
+		$scheduleTime = $this->parseTimestamp($input->getOption('schedule-time'));
+		$deleteTime = $this->parseTimestamp($input->getOption('delete-time'));
+
+		// validation
+		if ($scheduleTime && $deleteTime && $deleteTime < $scheduleTime) {
+			throw new \InvalidArgumentException("Publishing time is after deletion time");
```
Comment (nickvergessen) [inline lib/Command/Announce.php:146]:
```suggestion
			$output->writeln('Publishing time is after deletion time');
			return 2;
```

File: lib/Command/Announce.php [hunk 7]
```diff
@@ -0,0 +1,197 @@
...
+		}
+
+		$plainMessage = $this->plainifyMessage($message);
+		$notificationOptions = $this->notificationType->setNotificationTypes($activities, $notifications, $emails);
+
+		if (!$notificationOptions) {
+			throw new \InvalidArgumentException("You didn't set any notification option, please set 'activities', 'notifications' or 'emails'");
```
Comment (nickvergessen) [inline lib/Command/Announce.php:153]: This is not required in the web?

File: lib/Command/AnnouncementDelete.php [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+			->setDescription('Delete announcement by id')
+			->addArgument(
+				'id',
+				InputArgument::REQUIRED,
+				'Id of announcement to delete',
+			);
+	}
+
+	protected function execute(InputInterface $input, OutputInterface $output): int {
+		$deleteId = $this->parseId($input->getArgument('id'));
```
Comment (nickvergessen) [inline lib/Command/AnnouncementDelete.php:55]: should catch the thrown exception and writeln + return > 0

File: lib/Command/AnnouncementList.php [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+				InputArgument::OPTIONAL,
+				'Maximal number of announcements listed',
+				10,
+			);
+	}
+
+	protected function execute(InputInterface $input, OutputInterface $output): int {
+		$ulimit = $input->getArgument('limit');
+		if (!is_numeric($ulimit)) {
+			throw new \InvalidArgumentException('"' . $ulimit . '" is not numeric');
```
Comment (nickvergessen) [inline lib/Command/AnnouncementList.php:54]: writeln + return > 0

File: lib/Command/Announce.php (hunk 4 above)
Comment (mwinkens) [reply lib/Command/Announce.php:86]: That's some rough denglish :laughing:

File: lib/Command/Announce.php (hunk 7 above)
Comment (mwinkens) [reply lib/Command/Announce.php:155]: Apprently not, your message just gets ignored: https://github.com/nextcloud/announcementcenter/blob/92c90218fdc7fa3db938ddeb48c50bcb882d1d37/lib/Manager.php#L149

**Comment by mwinkens** [comment]
@nickvergessen ready for review again :tada:

File: lib/Command/Announce.php (hunk 7 above)
Comment (nickvergessen) [reply lib/Command/Announce.php]: Well it's still posted and can be read from the announcementcenter page

Comment (mwinkens) [reply lib/Command/Announce.php]: Oh! This is how this works, okay I'll remove this restriction

File: lib/Command/AnnouncementList.php [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+
+			$text = $this->formatTableRow($texts, $widths);
+			$output->writeln($text);
+		}
+		return 0;
+	}
+
+	private function ellipseAndPadText(string $text, int $width, string $sep = " "): string {
+		$text = str_pad($text, $width, $sep, STR_PAD_RIGHT);
+		$text = strlen($text) > $width ? substr($text, 0, $width - 3) . "..." : $text;
```
Comment (nickvergessen) [inline lib/Command/AnnouncementList.php:89]: Should also trim on new-lines or replace them with spaces, so the table doesn't break? (ignore my attempts figuring out the line breaks)
grafik


```suggestion
		$text = strlen($text) > $width ? substr($text, 0, $width - 2) . " …" : $text;
```

Comment (mwinkens) [reply lib/Command/AnnouncementList.php]: Never thought of line breaks, nice catch!

Comment (nickvergessen) [reply lib/Command/AnnouncementList.php]: Something along:
```
$text = str_replace("\n", ' ', $text);
```

Should work

Comment (mwinkens) [reply lib/Command/AnnouncementList.php]: close :+1: `str_replace(array("\r", "\n"), ' ', $text);`

**Review by nickvergessen — APPROVED** [review]

**Comment by pierreozoux** [comment]
Really nice!

Thanks @nickvergessen and @mwinkens \o/

**Pull request merged**